game/hud/src/widgets/Warband/services/session/layout.ts

```diff
@@ -56,35 +61,94 @@ export function savePosition(name: string, pos: Position): LayoutAction {
 
 // async
 
-let defaultWidgets = () => {
+const defaultWidgets = () : any => {
+  // this needs to return positions as % based on 1440x900 (default window size)
+  const w: number = 1440, h: number = 900;
+  alert('defaultWidgets widgets/Warbadn/services/session');
   return {
-    'PlayerHealth': {x:100,y:300,scale:1},
-    'TargetHealth': {x:600,y:300,scale:1},
-    'Party': {x:100,y:10,scale:1},
+    'PlayerHealth': { x: 100/w, y: 300/h, height: 200/h, width: 350/w, scale:1 },
+    'TargetHealth': { x: 600/w, y: 300/h, height: 200/h, width: 350/w,  scale:1 },
+    'Warband': { x: 100/w, y: 10/h , scale:1, height: 200/h, width: 350/w },
+    'Party': { x: 100/w, y: 10/h, scale:1, height: 200/h, width: 350/w },
   }
 }
 
 const minScale = 0.25;
 
 export interface LayoutState {
+  version?: number;
   locked?: boolean;
   widgets: any; // dictionary<name, position>
 }
 
-const initialState = {
+const initialState: LayoutState = {
+  version: CURRENT_STATE_VERSION,
   locked: true,
   widgets: defaultWidgets(),
 }
 
-let getInitialState = () => {
-  var storedState = JSON.parse(localStorage.getItem(localStorageKey)) as LayoutState;
+interface Size {
+  width: number;
+  height: number;
+}
+
+function px2pcnt(pos: Position, screen: Size) : Position {
+  // converts pixel co-ordinates into % of display size
+  return {
+    x: pos.x / screen.width,
+    y: pos.y / screen.height,
+    width: pos.width / screen.width,
+    height: pos.height / screen.height,
+    scale: 1
```

Comment:
If this was an exported method, I would agree, it's not though, it is private, and we know where the input values are coming from.  If there were some sensible defaults to apply, then that would be one thing but there are not.

In testing even a minimized window had an innerWidth/Height (screen size) of 1x1 pixels in the browser, so divide by zero is not going to happen.  If things do go wrong and they are undefined or zero and exception with be thrown which will abort the save, which is actually exactly what we want, because the only parameter checking I can sensible do is to check for the zero/undefined case and throw an exception.